### PR TITLE
changelog: scheduler plugins changes are API changes, not upgrade action required

### PR DESCRIPTION
**Pull request by BenTheElder**
also, reformatted to present these notes together instead of NodeToStatusMap change reading without context

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

I was reviewing the proposed release notes, these notes are confusing (the NodeStatusMap change doesn't mention scheduler plugins, looks like an attempt at indenting as a sub-bullet that didn't render as expected) **and are not upgrade action required, they're API changes**.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @kubernetes/release-managers, I know it's late in 1.32 but PTAL.

cc @macsko @googs1025

**Comment by liggitt** [comment]
/lgtm
/approve

will let release managers indicate how best to incorporate this for 1.32.0

**Comment by satyampsoni** [comment]
Ok! I am putting this change under API changes in the final release notes draft.

**Comment by googs1025** [comment]
/lgtm  for `PodEligibleToPreemptOthers` part.

cc @sanposhiho Please confirm this as well. :)

**Review by sanposhiho — APPROVED** [review]
/lgtm

How is the position in the release note determined? Should we add `api-change` label whenever we made this kind of implementation-level breaking changes for custom plugins, and then the note would go to the correct place?

**Comment by npolshakova** [comment]
> /lgtm
> 
> How is the position in the release note determined? Should we add `api-change` label whenever we made this kind of implementation-level breaking changes for custom plugins, and then the note would go to the correct place?

Yep, the release-notes are gathered and generated with the [krel](https://github.com/kubernetes/release/tree/master/cmd/release-notes) tool. I think it places it based on the label (ie. adding `kind api-change`) on the PR.

**Review by npolshakova — APPROVED** [review]

**Comment by BenTheElder** [comment]
> How is the position in the release note determined? Should we add api-change label whenever we made this kind of implementation-level breaking changes for custom plugins, and then the note would go to the correct place?

I'll defer to @liggitt on that one, I think the label may also be used for api-reviews which nominally (?) cover the REST API as opposed to go changes. But that category seemed more adequate than the "you have to read this before upgrading your cluster" section at the top we use for major breaking changes.

It might be that we should have a category for go API changes versus REST API?
I think that's usually covered by the go API being staged and handling it in the staging repo? (like https://github.com/kubernetes/client-go/blob/master/CHANGELOG.md)

**Comment by BenTheElder** [comment]
aside: it seems we didn't get the 1.32 docs lead in changelog/OWNERS unlike past releases?
https://github.com/kubernetes/kubernetes/blob/a499facee693a1a83daadb82d88f7b51d324ffc5/CHANGELOG/OWNERS#L6-L16

https://github.com/kubernetes/sig-release/blob/master/releases/release-1.32/release-team.md

cc @fsmunoz

File: CHANGELOG/CHANGELOG-1.32.md
Comment (sftim) [inline CHANGELOG/CHANGELOG-1.32.md:1071]: Maybe drop this heading, then?

Comment (xmudrii) [reply CHANGELOG/CHANGELOG-1.32.md:1071]: `## Urgent Upgrade Notes` should be removed as well

File: CHANGELOG/CHANGELOG-1.32.md
Comment (sftim) [inline CHANGELOG/CHANGELOG-1.32.md:1081]:
```suggestion
- **ACTION REQUIRED** for custom scheduler plugin developers:
```

**Review by sftim — COMMENTED** [review]
One more optional tweak

File: CHANGELOG/CHANGELOG-1.32.md
Comment (BenTheElder) [reply CHANGELOG/CHANGELOG-1.32.md:1071]: SGTM, I wasn't sure about that with the standardized layout. You could argue that it's more or less re-assuring to have the section empty vs missing and I didn't think a lot about that yet.

Dropping.

**Comment by BenTheElder** [comment]
made suggested changes

**Review by sanposhiho — APPROVED** [review]
/lgtm

**Comment by sanposhiho** [comment]
> It might be that we should have a category for go API changes versus REST API?

Yeah, maybe. For the scheduler, there are many interfaces and types that are exposed for the custom plugins and they're tightly coupled to the scheduler (of course). So, we often cannot avoid making this kind of implementation-level breaking changes that custom plugin developers will have to deal with at their upgrade. 
We can put them on the api-change section if that's the instruction from the release team though, those Go API changes don't interest most of users, but are only important for custom plugin developers; So, they might be just a noise for users who just use Kubernetes.

I remember, in the past, there was a proposal to move the scheduling framework onto the staging repo, but I'm, personally, not sure if that's possible honestly; at least, that would need a big effort, I guess.

**Comment by BenTheElder** [comment]
> We can put them on the api-change section if that's the instruction from the release team though, those Go API changes don't interest most of users, but are only important for custom plugin developers; So, they might be just a noise for users who just use Kubernetes.

Well, I'm not saying they're not important personally, just that they're a different audience than the "you can't upgrade your cluster without reading this" heading aimed at cluster operators. I think the API changes section is OK.


----

If there's somewhere more appropriate to follow-up, I'd move the rest of this discussion about the scheduler framework there and just the release notes changes here 😅 

That said, couple thoughts so I don't forget ...

<details> 

> Yeah, maybe. For the scheduler, there are many interfaces, types, functions etc that are exposed for the custom plugins and they're tightly coupled to the scheduler (of course). So, we often cannot avoid making this kind of implementation-level breaking changes that custom plugin developers will have to deal with at their upgrade.
>
> [...]
>
> I remember, in the past, there was a proposal to move the scheduling framework onto the staging repo, but I'm, personally, not sure if that's possible honestly; at least, that would need a big effort, I guess.



We could expose a scheduling plugin interface that is versioned without staging the entire scheduler?
It seems reasonable to commit to a set of hooks for scheduling behavior without exposing _all_ of the internals, and to publish that as a library or API.

As a general rule, the rest of the project states "importing kubernetes/kubernetes IS NOT SUPPORTED", there are many threads with this, and I'm kinda surprised that https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/624-scheduling-framework/README.md#custom-scheduler-plugins-out-of-tree didn't create more discussion.
The main reason we have staging is to carve out bits that aren't developed as proper external libraries but are supported to import. 

Also  ... has an RPC based approach with a versioned interface been discussed? (like CRI-API)

The scheduler might be the only place in the core project where we actually support "import the binary directly and extend it" instead of RPC / exec / IPC type extension mechanisms.

</details>

**Comment by sanposhiho** [comment]
> It seems reasonable to commit to a set of hooks for scheduling behavior without exposing all of the internals, and to publish that as a library or API.

Yeah, I know it's reasonable, just saying it's a big effort and hence no one is willing to promote until now. I guess there is (or was if it's closed by the bot) an issue discussing it. 

> Also ... has an RPC based approach with a versioned interface been discussed? (like CRI-API)

We have a feature called the extender, which is a webhook-style extension. 
But, that is not performant enough, especially for big clusters (and the performance is very critical factor for scheduling). Also, they have only a few feature sets, and it's not a simple problem that we can add more features onto it because of the performance issue; as you can see the scheduling framework, in order to support rich scheduling use cases, the scheduler and the plugin has to interact frequently, which would cause the latency issue more, if we just implemented them onto the extender.
Speaking of the history, actually the scheduling framework (the plugin system) came after the extender to solve the performance (+ other) problem. We actually don't put much effort into the extender especially after the scheduling framework arrives. The community's standard is also basically using the plugins because of the performance, rich features, etc.
But, we know the extender could be still valuable for users with simple use cases or small clusters since it's completely decoupled from Kubernetes scheduler (no need to rebuild etc at every version upgrade), and we're still maintaining it, also no plan to deprecate it for now.

Also, we have a newer side project [wasm-extension](https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension), which is more performant/richer feature-sets than the extender, while the extension is completely decoupled from Kubernetes binary.

**Comment by sftim** [comment]
Can we approve this prior to the v1.32 release?

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve
/milestone v1.32
/hold
cc @fsmunoz for final approval

**Comment by xmudrii** [comment]
@sftim It's now good to go, let's just make sure that the @kubernetes/release-team-leads are okay with merging this.

**Pull request merged**
